File: voicecal-service/src/main/java/com/voicecal/service/repository/ReminderRepository.java
package com.voicecal.service.repository;

import com.voicecal.core.model.Reminder;
import com.voicecal.core.model.ReminderStatus;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

/**
 * Spring Data JPA repository for Reminder entities.
 */
@Repository
public interface ReminderRepository extends JpaRepository<Reminder, UUID> {

    /**
     * Find all pending reminders that are due (remindAt is at or before the given time).
     */
    @Query("SELECT r FROM Reminder r " +
           "WHERE r.status = 'PENDING' " +
           "AND r.remindAt <= :now " +
           "ORDER BY r.remindAt ASC")
    List<Reminder> findDueReminders(@Param("now") LocalDateTime now);

    /**
     * Find all reminders for a specific event.
     */
    List<Reminder> findByEventId(UUID eventId);

    /**
     * Find all reminders for a specific event with a given status.
     */
    List<Reminder> findByEventIdAndStatus(UUID eventId, ReminderStatus status);

    /**
     * Find all pending reminders for a specific event.
     */
    @Query("SELECT r FROM Reminder r " +
           "WHERE r.eventId = :eventId " +
           "AND r.status = 'PENDING' " +
           "ORDER BY r.remindAt ASC")
    List<Reminder> findPendingRemindersByEventId(@Param("eventId") UUID eventId);

    /**
     * Delete all reminders associated with a specific event.
     */
    void deleteByEventId(UUID eventId);

    /**
     * Cancel all pending reminders for a specific event.
     */
    @Modifying(clearAutomatically = true, flushAutomatically = true)
    @Query("UPDATE Reminder r SET r.status = 'CANCELLED' " +
           "WHERE r.eventId = :eventId AND r.status = 'PENDING'")
    void cancelPendingRemindersByEventId(@Param("eventId") UUID eventId);
}

File: voicecal-mobile/src/services/database/ReminderRepository.ts
import { Reminder, ReminderStatus } from '../../models';
import { databaseService } from './DatabaseService';

const COLS = ['id', 'event_id', 'remind_at', 'status'];

function toRow(r: Reminder): Record<string, unknown> {
  return {
    id: r.id,
    event_id: r.eventId,
    remind_at: r.remindAt,
    status: r.status,
  };
}

function fromRow(row: Record<string, unknown>): Reminder {
  return {
    id: row.id as string,
    eventId: row.event_id as string,
    remindAt: row.remind_at as string,
    status: row.status as ReminderStatus,
  };
}

export class ReminderRepository {
  async save(reminder: Reminder): Promise<Reminder> {
    const db = await databaseService.getDatabase();
    const r = toRow(reminder);
    await db.runAsync(
      `INSERT OR REPLACE INTO reminders (${COLS.join(', ')}) VALUES (${COLS.map(() => '?').join(', ')})`,
      COLS.map((c) => r[c]) as (string | number | null)[]
    );
    return reminder;
  }

  async findDueReminders(now: string): Promise<Reminder[]> {
    const db = await databaseService.getDatabase();
    const rows = await db.getAllAsync(
      `SELECT ${COLS.join(', ')} FROM reminders WHERE status = ? AND remind_at <= ? ORDER BY remind_at ASC`,
      [ReminderStatus.PENDING, now]
    );
    return (rows as Record<string, unknown>[]).map(fromRow);
  }

  async findByEventId(eventId: string): Promise<Reminder[]> {
    const db = await databaseService.getDatabase();
    const rows = await db.getAllAsync(
      `SELECT ${COLS.join(', ')} FROM reminders WHERE event_id = ? ORDER BY remind_at ASC`,
      [eventId]
    );
    return (rows as Record<string, unknown>[]).map(fromRow);
  }

  async updateStatus(id: string, status: ReminderStatus): Promise<void> {
    const db = await databaseService.getDatabase();
    await db.runAsync('UPDATE reminders SET status = ? WHERE id = ?', [status, id]);
  }

  async deleteByEventId(eventId: string): Promise<void> {
    const db = await databaseService.getDatabase();
    await db.runAsync('DELETE FROM reminders WHERE event_id = ?', [eventId]);
  }

  async cancelPendingByEventId(eventId: string): Promise<void> {
    const db = await databaseService.getDatabase();
    await db.runAsync(
      "UPDATE reminders SET status = ? WHERE event_id = ? AND status = 'PENDING'",
      [ReminderStatus.CANCELLED, eventId]
    );
  }
}

export const reminderRepository = new ReminderRepository();
